application/src/main/java/org/thingsboard/server/service/transport/DefaultTransportApiService.java

```diff
@@ -665,4 +727,66 @@ private ListenableFuture<TransportApiResponseMsg> queuesToTransportApiResponseMs
+        if (matcher.find()) {
+            return matcher.group(0);
+        }
+        return null;
+    }
+
+    private List<String> convertX509CertificateChainToList(String certificateChain) {
+        List<String> chain = new ArrayList<>();
+        String regex = "-----BEGIN CERTIFICATE-----\\s*.*?\\s*-----END CERTIFICATE-----";
+        Pattern pattern = Pattern.compile(regex);
+        Matcher matcher = pattern.matcher(certificateChain);
+        while (matcher.find()) {
+            chain.add(EncryptionUtil.certTrimNewLines(matcher.group()));
+        }
+        return chain;
+    }
+
+    private X509Certificate readCertFile(String fileContent) throws Exception {
+        X509Certificate certificate = null;
+        if (fileContent != null && !fileContent.trim().isEmpty()) {
+            fileContent = fileContent.replace("-----BEGIN CERTIFICATE-----", "")
+                    .replace("-----END CERTIFICATE-----", "")
+                    .replaceAll("\\s", "");
+            byte[] decoded = Base64.decodeBase64(fileContent);
+            CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+            try (InputStream inStream = new ByteArrayInputStream(decoded)) {
+                certificate = (X509Certificate) certFactory.generateCertificate(inStream);
+            }
+        }
+        return certificate;
+    }
+
+    private DeviceCredentials updateDeviceCredentials(TenantId tenantId, DeviceCredentials deviceCredentials, String certificateValue,
+                                                      String certificateHash, DeviceCredentialsType credentialsType) {
+        deviceCredentials.setCredentialsId(certificateHash);
+        deviceCredentials.setCredentialsValue(certificateValue);
+        deviceCredentials.setCredentialsType(credentialsType);
+        return deviceCredentialsService.updateDeviceCredentials(tenantId, deviceCredentials);
+    }
+
+    private DeviceCredentials createDeviceCredentials(TenantId tenantId, DeviceId deviceId, String certificateValue,
+                                                      String certificateHash, DeviceCredentialsType credentialsType) {
+        DeviceCredentials createDevCredentials = new DeviceCredentials();
+        createDevCredentials.setDeviceId(deviceId);
+        createDevCredentials.setCredentialsType(credentialsType);
+        createDevCredentials.setCredentialsId(certificateHash);
+        createDevCredentials.setCredentialsValue(certificateValue);
+        return deviceCredentialsService.createDeviceCredentials(tenantId, createDevCredentials);
+    }
+
+    private Device createDevice(TenantId tenantId, DeviceProfileId deviceProfileId, String deviceName, String type) {
+        Device device = new Device();
+        device.setTenantId(tenantId);
+        device.setDeviceProfileId(deviceProfileId);
+        device.setName(deviceName);
+        device.setType(type);
+        return deviceService.saveDevice(device);
```

Comment:
Please use TbDeviceService from application layer, instead of DeviceService from DAO layer - this will provide correct audit logging, notification of other services, etc.